internal/pkg/propagation/stdlib.go

```diff
@@ -0,0 +1,154 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package propagation
+
+import (
+	"go/types"
+	"strings"
+
+	"github.com/google/go-flow-levee/internal/pkg/utils"
+	"golang.org/x/tools/go/ssa"
+)
+
+// taintStdlibCall propagates taint through a static call to a standard
+// library function, provided that the function's taint propagation behavior
+// is known (i.e. the function has a summary).
+func (prop *Propagation) taintStdlibCall(call *ssa.Call, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
+	summ, ok := funcSummaries[funcName(call)]
+	if !ok {
+		return
+	}
+
+	prop.taintFromSummary(summ, call, call.Call.Args, maxInstrReached, lastBlockVisited)
+}
+
+// taintStdlibInterfaceCall propagates taint through a call to a function that
+// implements an interface function from the standard library, provided that the
+// function's taint propagation behavior is known (i.e. the function has a summary).
+// This can be a static call or a dynamic call.
+func (prop *Propagation) taintStdlibInterfaceCall(call *ssa.Call, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
+	summ, ok := interfaceFuncSummaries[funcKey{funcNameWithoutReceiver(call), sigTypeString(call.Call.Signature())}]
+	if !ok {
+		return
+	}
+
+	var args []ssa.Value
+	// For "invoke" calls, Value is the receiver
+	if call.Call.IsInvoke() {
+		args = append(args, call.Call.Value)
+	}
+	args = append(args, call.Call.Args...)
+
+	prop.taintFromSummary(summ, call, args, maxInstrReached, lastBlockVisited)
+}
+
+func (prop *Propagation) taintFromSummary(summ summary, call *ssa.Call, args []ssa.Value, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
+	tainted := int64(0)
+	for i, a := range args {
+		if prop.tainted[a.(ssa.Node)] {
+			tainted |= 1 << i
+		}
+	}
+	if (tainted & summ.ifTainted) == 0 {
+		return
+	}
+
+	for _, i := range summ.taintedArgs {
+		prop.taint(args[i].(ssa.Node), maxInstrReached, lastBlockVisited, false)
+	}
+
+	if call.Referrers() == nil {
+		return
+	}
+
+	if call.Call.Signature().Results().Len() == 1 {
+		if len(summ.taintedRets) > 0 {
+			prop.taintReferrers(call, maxInstrReached, lastBlockVisited)
+		}
+		return
+	}
+
+	indexToExtract := map[int]*ssa.Extract{}
+	for _, r := range *call.Referrers() {
+		e := r.(*ssa.Extract)
+		indexToExtract[e.Index] = e
+	}
+	for i := range summ.taintedRets {
+		prop.taint(indexToExtract[i], maxInstrReached, lastBlockVisited, true)
+	}
```

Comment:
I think I follow this only because I am deeply familiar with `ssa`.  Comments throughout this method could be useful.  Then again, "deeply familiar with `ssa`" is rather the intended audience at this depth.

Reply:
Added some high-level comments.